System container support

codificat opened the pull request:
Add files and documentation to support running images built from playbook2image to be run as system containers.

One of the changes (avoid `/etc/passwd` update for `root`) is kind of independent, but it is specially relevant to this use case because normal usage is expected to run under non-root.

WIP because it's probably worth waiting for a release of `atomic` that includes https://github.com/projectatomic/atomic/pull/952 and to add the corresponding label, test and update the docs. Feedback is welcome meanwhile :)

codificat committed:
Avoid /etc/passwd update when run as root

The update to the passwd file is not needed when running as root because there's
already a correct entry for UID 0 in /etc/passwd. Moreover, if this is run with
a read-only filesystem (for example the system container use case does this) it
causes the 'run' script to fail.

codificat committed:
System container support

Add files and documentation to support running images built from playbook2image
to be run as system containers.

aweiteka commented on system-container/README.md:
```diff
@@ -0,0 +1,60 @@
+# System Container support
+
+A system container runs using runC instead of Docker and it is managed by the
```
Maybe better to say the process is managed by systemd and installed using the atomic tool.

codificat replied:
I believe systemd is not involved in the "run once" scenario

aweiteka commented on system-container/README.md:
```diff
@@ -0,0 +1,60 @@
+to run as a system container, so no specific action needs to be taken to build a
+system container-ready image based on playbook2image.
+
+Once a container image is built (e.g. `my/image`) you can import it into the OSTree storage:
```
Let's remind or clarify that this build step is where the user adds their playbooks to the base image.

codificat replied:
Right, will clean up this a bit more - I took this content from contributed docs to the openshift-ansible repo and it's indeed worth customizing a bit more for playbook2image...

aweiteka commented on system-container/README.md:
```diff
@@ -0,0 +1,60 @@
+
+    atomic uninstall myimage
+
+*NOTE:* after the next release of the `atomic` command that includes [PR 952](https://github.com/projectatomic/atomic/pull/952) we will add an `atomic.run=once` label and the `systemctl` invocation or the cleanup/uninstall won't be necessary.
```
PR#952 has been merged.

codificat replied:
Yes it was, but just recently (last week) there was a release of the `atomic` tool with the changes.

Now that this is actually available I will do some actual testing :) and update

aweiteka commented on system-container/README.md:
```diff
@@ -0,0 +1,60 @@
+
+*NOTE:* after the next release of the `atomic` command that includes [PR 952](https://github.com/projectatomic/atomic/pull/952) we will add an `atomic.run=once` label and the `systemctl` invocation or the cleanup/uninstall won't be necessary.
+
+If the playbooks need additional files then they can use the path `/var/lib/playbook2image` in the container as it is bind mounted from the host (the source directory is controllable with `VAR_LIB_PLAYBOOK2IMAGE`).
```
Is there a default host path you can call out here?

aweiteka reviewed (changes requested):
Great to see this capability added. Can you mention the use case in this README? Why would I choose to use a system container vs running a container manually?

codificat committed:
Add atomic run once label

After projectatomic/atomic#952 it's now possible to add a label to an image
atomic.run=once so that when installed as a system container it will be run
immediately with runc (no systemd involved) and then uninstalled.

This use case matches containerized Ansible playbooks very well, so we add
that label in playbook2image so that the usage experience of packaged
playbooks run as system containers matches standalone container runs.

codificat commented:
@aweiteka I expanded the docs a bit more with your feedback.

Also added the `atomic.run="once"` label for the system container use case. I believe that for system containers coming from playbook2image we will always want to use that *run once* approach. What do you think? Is there a use case where we'd prefer systemd to manage the execution of the playbook instead? If so, should we then document how to remove that label in the `docker build`-based docs, or should we instead not add the label by default and document how to add it?

aweiteka commented:
> What do you think? Is there a use case where we'd prefer systemd to manage the execution of the playbook instead?

I think it's fine as-is right now. We can get this in and get user feedback.

Is it okay with you to merge?

codificat commented:
LGTM, ship it!
